spring-cloud-gcp-samples/spring-cloud-gcp-data-firestore-sample/src/test/java/com/example/NativeSupportIntegrationTest.java

```diff
@@ -0,0 +1,158 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.RestTemplate;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@Testcontainers
+@Tag("native")
+public class NativeSupportIntegrationTest {
+
+	private static final String TESTER_NAME = "Tester";
+	private static final int TESTER_AGE = 42;
+	private static final String PET1_TYPE = "Dinosaur";
+	private static final String PET1_NAME = "Dino";
+	private static final String PET2_TYPE = "Project";
+	private static final String PET2_NAME = "LoCoRepo";
+	private static final String MY_GCP_PROJECT_ID = "my-gcp-project-id";
+
+	@Container
+	private static final FirestoreEmulatorContainer emulator = new FirestoreEmulatorContainer(
+			DockerImageName.parse("gcr.io/google.com/cloudsdktool/cloud-sdk:316.0.0-emulators")
+	);
+	@Test
+	void testApplicationNativeDockerImage() throws ExecutionException, InterruptedException {
+		Integer firestorePort = emulator.getMappedPort(8080);
+		org.testcontainers.Testcontainers.exposeHostPorts(firestorePort);
+
+		MyAppNativeDockerContainer myApp = new MyAppNativeDockerContainer(emulator);
+		try {
+			myApp.start();
+			RestTemplate restTemplate = new RestTemplate();
+
+			HttpHeaders headers = new HttpHeaders();
+			headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
+
+			MultiValueMap<String, String> map = new LinkedMultiValueMap<>();
+			map.add("name", TESTER_NAME);
+			map.add("age", String.valueOf(TESTER_AGE));
+			map.add("pets", String.format("%s-%s,%s-%s", PET1_TYPE, PET1_NAME, PET2_TYPE, PET2_NAME));
+
+			HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(map, headers);
+
+
+			String baseUsersUrl = myApp.getUsersUrl() + "/users";
+			ResponseEntity<User> response = restTemplate.postForEntity(baseUsersUrl + "/saveUser", request, User.class);
+			assertThat(response.getStatusCode())
+					.isEqualTo(HttpStatus.OK);
+			User expectedUser = new User(TESTER_NAME, TESTER_AGE, Arrays.asList(
+					new Pet(PET1_TYPE, PET1_NAME),
+					new Pet(PET2_TYPE, PET2_NAME)
+			));
+			assertThat(response.getBody())
+					.isNotNull()
+					.isEqualTo(expectedUser);
+
+			ResponseEntity<User[]> listUsersResp = restTemplate.getForEntity(baseUsersUrl, User[].class);
+			assertThat(listUsersResp.getStatusCode())
+					.isEqualTo(HttpStatus.OK);
+			assertThat(listUsersResp.getBody())
+					.hasSize(1)
+					.containsExactly(expectedUser);
+
+			FirestoreOptions options = FirestoreOptions.getDefaultInstance().toBuilder()
+					.setHost(emulator.getEmulatorEndpoint())
+					.setCredentials(NoCredentials.getInstance())
+					.setProjectId(MY_GCP_PROJECT_ID)
+					.build();
+			Firestore firestore = options.getService();
+
+			CollectionReference users = firestore.collection("user");
```

Comment:
add these tests made me realize the document name is somehow not working...

Reply:
I just double checked and it's working. Maybe a SNAPSHOT version of dependencies was downloaded when I was testing and since that didn't have my changes, the test failed, but I verified it's not happening with "fresh build".